File: cotta-guice/src/main/java/com/mgtriffid/games/cotta/experimental/guice/serialization/InputSerialization.java
package com.mgtriffid.games.cotta.experimental.guice.serialization;

import com.mgtriffid.games.cotta.experimental.guice.serialization.recipe.InputRecipe;

public interface InputSerialization {
    byte[] serializeInputRecipe(InputRecipe recipe);
    InputRecipe deserializeInputRecipe(byte[] bytes);
}

File: cotta-core/src/main/kotlin/com/mgtriffid/games/cotta/core/serialization/InputSerialization.kt
package com.mgtriffid.games.cotta.core.serialization

import com.mgtriffid.games.cotta.core.entities.PlayerId
import com.mgtriffid.games.cotta.core.input.PlayerInput

interface InputSerialization<IR: InputRecipe> {
    fun serializeInputRecipe(recipe: IR): ByteArray
    fun serializeInput(input: PlayerInput): ByteArray
    fun deserializeInputRecipe(bytes: ByteArray): IR
    fun deserializeInput(bytes: ByteArray): PlayerInput
    fun serializePlayersInputs(inputs: Map<PlayerId, PlayerInput>): ByteArray
    fun deserializePlayersInputs(bytes: ByteArray): Map<PlayerId, PlayerInput>
}
